### ISSUE TITLE
[ksbeattie on 2007-03-12 17:45:13] : DAQLaunch.py randomly doesn't return

### ISSUE THREAD

**Issue by dglo** [dglo/pdaq#184]
No associated GitHub commit

**Comment by dglo** [comment]
[ksbeattie on 2007-03-14 00:13:51]
Investigating this, it appears that the read at line 30 of ParallelShell.py is intermittently blocking, causing the done var to never get set to True in the following line.

What is very odd if that after instrumenting this run_thead() and the ParallelShell.wait methods, when that read is blocked, if I kill the component which wait says it is waiting on, it doesn't return from the read in run_thread.  Killing a different process *does* cause it to return though!  Something is getting hosed here somewhere beyond a read getting blocked...

The implementation of commands.getstatusoutput() is nearly the same as is done in run_thread, so the read blocking must be because: 1) the command is not returning anything on stdout or stderr (which makes sense as it is redirected, I believe) and 2) the command is not returning - why I'm not sure of yet...\n\n

**Comment by dglo** [comment]
[ksbeattie on 2007-03-14 00:10:33]
Another clue: When DAQLaunch doesn't return the following appears in the catchall.log file:

Catchall org.apache.xmlrpc.server.XmlRpcStreamServer-XML-RPC-0 ERROR [Wed Mar 14 00:06:19 PDT 2007] execute: Error while performing request
Catchall > org.apache.xmlrpc.XmlRpcException: Failed to invoke method getState in class icecube.daq.juggler.component.DAQCompServer: Component#8 not found
Catchall >      at org.apache.xmlrpc.server.ReflectiveXmlRpcHandler.invoke(ReflectiveXmlRpcHandler.java:122)
Catchall >      at org.apache.xmlrpc.server.ReflectiveXmlRpcHandler.execute(ReflectiveXmlRpcHandler.java:102)
Catchall >      at org.apache.xmlrpc.server.XmlRpcServerWorker.execute(XmlRpcServerWorker.java:43)
Catchall >      at org.apache.xmlrpc.server.XmlRpcServer.execute(XmlRpcServer.java:83)
Catchall >      at org.apache.xmlrpc.server.XmlRpcStreamServer.execute(XmlRpcStreamServer.java:182)
Catchall >      at org.apache.xmlrpc.webserver.Connection.run(Connection.java:175)
Catchall >      at org.apache.xmlrpc.util.ThreadPool$MyThread.runTask(ThreadPool.java:71)
Catchall >      at org.apache.xmlrpc.util.ThreadPool$MyThread.run(ThreadPool.java:87)
Catchall > Caused by:
Catchall > icecube.daq.juggler.component.DAQCompException: Component#8 not found
Catchall >      at icecube.daq.juggler.component.DAQCompServer.getState(DAQCompServer.java:495)
Catchall >      at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Catchall >      at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
Catchall >      at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
Catchall >      at java.lang.reflect.Method.invoke(Method.java:585)
Catchall >      at org.apache.xmlrpc.server.ReflectiveXmlRpcHandler.invoke(ReflectiveXmlRpcHandler.java:111)
Catchall >      at org.apache.xmlrpc.server.ReflectiveXmlRpcHandler.execute(ReflectiveXmlRpcHandler.java:102)
Catchall >      at org.apache.xmlrpc.server.XmlRpcServerWorker.execute(XmlRpcServerWorker.java:43)
Catchall >      at org.apache.xmlrpc.server.XmlRpcServer.execute(XmlRpcServer.java:83)
Catchall >      at org.apache.xmlrpc.server.XmlRpcStreamServer.execute(XmlRpcStreamServer.java:182)
Catchall >      at org.apache.xmlrpc.webserver.Connection.run(Connection.java:175)
Catchall >      at org.apache.xmlrpc.util.ThreadPool$MyThread.runTask(ThreadPool.java:71)
Catchall >      at org.apache.xmlrpc.util.ThreadPool$MyThread.run(ThreadPool.java:87)

I've actually seen more than one component 'fail' in this way at a time.

**Comment by dglo** [comment]
[ksbeattie on 2007-03-14 17:59:37]
more info:

ksb@moraine[pDAQ_trunk] 17:57:11 (0)$ uname -a
Linux moraine.lbl.gov 2.4.21-47.ELsmp #1 SMP Wed Jul 5 20:38:41 EDT 2006 i686 i686 i386 GNU/Linux

ksb@moraine[pDAQ_trunk] 17:57:13 (0)$ python
Python 2.5 (r25:51908, Nov 30 2006, 11:01:17) 
[GCC 3.2.3 20030502 (Red Hat Linux 3.2.3-56)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> 

Tracing DAQLaunch when it is not returning:
ksb@moraine[pDAQ_trunk] 17:56:25 (0)$ strace -p 15215
Process 15215 attached - interrupt to quit
select(0, NULL, NULL, NULL, {0, 30000}) = 0 (Timeout)
futex(0x95ac1b8, FUTEX_WAKE, 1)         = 0
select(0, NULL, NULL, NULL, {0, 250000}) = 0 (Timeout)
futex(0x95ac1b8, FUTEX_WAKE, 1)         = 0
select(0, NULL, NULL, NULL, {0, 250000}) = 0 (Timeout)
futex(0x95ac1b8, FUTEX_WAKE, 1)         = 0
select(0, NULL, NULL, NULL, {0, 250000}) = 0 (Timeout)
futex(0x95ac1b8, FUTEX_WAKE, 1)         = 0
futex(0x95ac1b8, FUTEX_WAKE, 1)         = 0
select(0, NULL, NULL, NULL, {0, 250000}) = 0 (Timeout)
futex(0x95ac1b8, FUTEX_WAKE, 1)         = 0

this goes on until I ctrl-c

select(0, NULL, NULL, NULL, {0, 250000}) = 0 (Timeout)
futex(0x95ac1b8, FUTEX_WAKE, 1)         = 0
select(0, NULL, NULL, NULL, {0, 250000}) = 0 (Timeout)
futex(0x95ac1b8, FUTEX_WAKE, 1)         = 0
select(0, NULL, NULL, NULL, {0, 250000} <unfinished ...>
Process 15215 detached\n\n

**Comment by dglo** [comment]
[jacobsen on 2007-03-26 08:48:03]
This won't get fixed in time for exmoor.

**Comment by dglo** [comment]
[dglo on 2007-03-26 16:19:57]
I spent some time on this today.  What's happening is that ParallelShell is running a bunch of threads which each start a component using os.popen() and then read the popen file handle to determine that the component has started.  For some reason, one or more of these threads fails to return, even though the corresponding StartComponent.py script has exited.

Given Keith's note about killing a different process to get the original thread to return, this smells like either a race condition inside Python or maybe popen() isn't thread-safe?

I just got a DAQLaunch to hang waiting for 1 of 5 StringHubs to start, and I was able to kill all the non-hanging StringHub processes and the event builder process without the popen returning.  I then killed off the 3 remaining processes and the hanging popen finally returned.  There's something very wrong here!\n\n

**Comment by dglo** [comment]
[ksbeattie on 2007-04-17 14:47:37]
Author: ksb
Date: 2007-04-17 17:46:18 -0400 (Tue, 17 Apr 2007)
New Revision: 1213

Added:
   projects/cluster-config/trunk/subprocess.py
Modified:
   projects/cluster-config/trunk/DeployPDAQ.py
   projects/cluster-config/trunk/ParallelShell.py
   projects/dash/trunk/DAQLaunch.py
Log:
Issue #1239: DAQLaunch.py randomly doesn't return

Re-worked ParallelShell.py to not use threads and rather use
subprocess.py which I've plucked from python 2.4 (it promised to be
backwards compatible to python 2.2).  ParallelShell.py has *slightly*
different semantics but for the most part are identical to it's
previous behavior. DAQLaunch.py required a few small changes for this
and DeployPDAQ.py got a bit more, but have essentially the same interface.
